types/test.tsx

```diff
@@ -0,0 +1,314 @@
+/* tslint-disable */
+// TypeScript Version: 2.7
+
+import * as React from 'react';
+import {
+  Active,
+  Input,
+  Counter,
+  Focus,
+  FocusManager,
+  Form,
+  Hover,
+  List,
+  Map,
+  Set,
+  State,
+  Toggle,
+  Touch,
+  Value,
+  ListRender,
+  SetRender,
+  MapRender,
+  StateRender,
+  ToggleRender,
+  TouchRender,
+  ValueRender,
+  ActiveRender,
+  InputRender, CounterRender, FocusRender, FocusManagerRender, FormRender, HoverRender,
```

Comment:
1. TS is not about "infering". It partially available in latest version.
2. "export everything you could export" is actually a DTSLint rule. I was also a bit surprised.
https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md

And it is not configurable.

Reply:
`no-relative-import-in-test` is actually a rule for DefinitelyTyped, and could not be used in this case. If you will change import as it asks - it would not work, at all.

Reply:
A lot of ts fans told me every time that type inference is good enough in typescript and I can remove flow. Are they liars?

Maybe remove dtslint? I'm not sure all these rules fits to out needs.

Reply:
I just think that types are part of module design and should be written carefully. If something is not required it should stay private.

Reply:
It works, but differently. 
If you will place render function into the “restricted” place - ie use as a render prop - arguments will be inferred. 
If you will describe them separately - you will have to type function. 

I will try to clarify “the state of export” as long all other project I maintain follow “don’t export internal shit” rules, and this is “obsolete” way, as long declaring “module name” in the beginning of a file. 
It case of making TS library and asking TS to generate imports for you - it will generate the same code we got here. 
Modern TS :shrug:

Reply:
It works, but differently. 
If you will place render function into the “restricted” place - ie use as a render prop - arguments will be inferred. 
If you will describe them separately - you will have to type function. 

I will try to clarify “the state of export” as long all other project I maintain follow “don’t export internal shit” rules, and this is “obsolete” way, as long declaring “module name” in the beginning of a file. 
It case of making TS library and asking TS to generate imports for you - it will generate the same code we got here. 
Modern TS :shrug:

Reply:
Just to add my 2 cents here.
I find myself wrapping lib components pretty frequently. Then you end up with types like this:

```ts
export type WrappedCompProps = MyProps & LibProps;
```

In that case, if the lib maintainer has decided to not export the lib props - you’re in a pretty bad position. Of course you could do smth like

```ts
type LibProps = LibComponent['props']
```
but that feels a little bit weird. If you only need to use the type of the attribute of a component prop, then it’s even uglier:

```ts
type RenderCallbackType = LibComponent['props']['children']
```

So exporting everything that has at least some possibility of being used by the consumer is generally the safest approach

Reply:
@TrySound if you don't think this is a good pattern - what is the alternative that you would propose?

Reply:
Use own data types for data and callbacks which wraps set or another transform function.

Reply:
I am wondering - what if move "index.d.ts" into "internals.d.ts", and "reexport" only chosen interfaces via "index", exporting the rest in a sub namespace.
All interfaces are still available, but "common" names of internal interfaces wont "jump" into autocomplete or auto import proposals.

Reply:
I think it's ok to export these *Render types for when needed, for example when you isolate the render in a separate function:

Reply:
I think it's ok to export these *Render types for when needed, for example when you isolate the render in a separate function:

Reply:
I think it's ok to export these *Render types for when needed, for example when you isolate the render in a separate function:

```jsx

Reply:
I think it's ok to export these *Render types for when needed, for example when you isolate the render in a separate function

```jsx
<Toggle render={({ on }) => () }>

Reply:
I think it's ok to export these *Render types for when needed, for example when you isolate the render in a separate function

```jsx
<Toggle render={({ on, togge }) => (
  /* TS knows that `on` is boolean here */
)}>
```

```jsx
const render = ({ on, toggle }) => (
  /* TS don't know what is `on` here */
)

<Toggle render={render} />
```

In this second case it is good that the person can type without any effort, and the best solution I see is to export the `Render` of each component.

```jsx
const render: ToggleRender = ({ on, toggle }) => (
  /* Now TS knows that `on` is boolean */
)

<Toggle render={render} />
```